### ISSUE TITLE
[Bug] Expert distribution recorder randomly drops recording for some tokens when batch size is larger than 32

### ISSUE THREAD

**Issue by KawaiiNotHawaii** [sgl-project/sglang#12823]
### Checklist

- [x] 1. I have searched related issues but cannot get the expected help.
- [x] 2. The bug has not been fixed in the latest version.
- [x] 3. Please note that if the bug-related issue you submitted lacks corresponding environment info and a minimal reproducible demo, it will be challenging for us to reproduce and resolve the issue, reducing the likelihood of receiving feedback.
- [x] 4. If the issue you raised is not a bug but a question, please raise a discussion at https://github.com/sgl-project/sglang/discussions/new/choose Otherwise, it will be closed.
- [x] 5. Please use English, otherwise it will be closed.

### Describe the bug

I'm training MoE model using verl + sglang (Version: 0.4.9.post6), and enabled the expert distribution recorder during training. 
I have an assertion on the number of non-zero attention masks and the number of token-level expert distribution matrix should be identical.
With bsz=32, rollout n=8 set in verl training config, the result can pass that assertion. But when I scale bsz to 64 or 128, it failed. Each rollout's number of non-zero attention mask is larger than the recorded token expert distribution and is by a fixed size (like 8) across all rollouts on that entry sglang engine.
Through my experiments, when an SGLang inference entry has a batch size larger than 32, it can happen.
### Reproduction

sglang Version: 0.4.9.post6
left padding prompt mask to max_prompt_length (e.g. 2048) and right padding response mask to max_response_length (e.g. 32*1024). then concatenate them to get the attention mask. Empirically, the non-zero locations are meaningful tokens whose token expert distribution should be recorded without any missing ones

### Environment

Python: 3.12.3 | packaged by Anaconda, Inc. | (main, May  6 2024, 19:46:43) [GCC 11.2.0]
CUDA available: True
GPU 0,1,2,3,4,5,6,7: NVIDIA H800
GPU 0,1,2,3,4,5,6,7 Compute Capability: 9.0
CUDA_HOME: /usr/local/cuda
NVCC: Cuda compilation tools, release 12.9, V12.9.41
CUDA Driver Version: 560.35.03
PyTorch: 2.7.1+cu126
sglang: 0.4.9.post6
sgl_kernel: 0.2.7
flashinfer_python: 0.2.9rc2
triton: 3.3.1
transformers: 4.54.0
torchao: 0.9.0
numpy: 1.26.4
aiohttp: 3.12.13
fastapi: 0.115.13
hf_transfer: 0.1.9
huggingface_hub: 0.36.0
interegular: 0.3.3
modelscope: 1.27.0
orjson: 3.10.18
outlines: 0.1.11
packaging: 25.0
psutil: 7.1.3
pydantic: 2.11.7
python-multipart: 0.0.20
pyzmq: 27.0.0
uvicorn: 0.34.3
uvloop: 0.21.0
vllm: Module Not Found
xgrammar: 0.1.21
openai: 1.88.0
tiktoken: 0.9.0
anthropic: 0.54.0
litellm: 1.72.6.post1
decord: 0.6.0
NVIDIA Topology: 
        GPU0    GPU1    GPU2    GPU3    GPU4    GPU5    GPU6    GPU7    NIC0    NIC1    NIC2    NIC3    NIC4    NIC5    NIC6    NIC7    NIC8    NIC9    CPU Affinity    NUMA Affinity   GPU NUMA ID
GPU0     X      NV18    NV18    NV18    NV18    NV18    NV18    NV18    PIX     PXB     NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     0-23,96-119     0               N/A
GPU1    NV18     X      NV18    NV18    NV18    NV18    NV18    NV18    PXB     PIX     NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     0-23,96-119     0               N/A
GPU2    NV18    NV18     X      NV18    NV18    NV18    NV18    NV18    NODE    NODE    PIX     PXB     NODE    SYS     SYS     SYS     SYS     SYS     0-23,96-119     0               N/A
GPU3    NV18    NV18    NV18     X      NV18    NV18    NV18    NV18    NODE    NODE    PXB     PIX     NODE    SYS     SYS     SYS     SYS     SYS     0-23,96-119     0               N/A
GPU4    NV18    NV18    NV18    NV18     X      NV18    NV18    NV18    SYS     SYS     SYS     SYS     SYS     PIX     PXB     NODE    NODE    NODE    48-71,144-167   2               N/A
GPU5    NV18    NV18    NV18    NV18    NV18     X      NV18    NV18    SYS     SYS     SYS     SYS     SYS     PXB     PIX     NODE    NODE    NODE    48-71,144-167   2               N/A
GPU6    NV18    NV18    NV18    NV18    NV18    NV18     X      NV18    SYS     SYS     SYS     SYS     SYS     NODE    NODE    PIX     PXB     NODE    48-71,144-167   2               N/A
GPU7    NV18    NV18    NV18    NV18    NV18    NV18    NV18     X      SYS     SYS     SYS     SYS     SYS     NODE    NODE    PXB     PIX     NODE    48-71,144-167   2               N/A
NIC0    PIX     PXB     NODE    NODE    SYS     SYS     SYS     SYS      X      PXB     NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS
NIC1    PXB     PIX     NODE    NODE    SYS     SYS     SYS     SYS     PXB      X      NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS
NIC2    NODE    NODE    PIX     PXB     SYS     SYS     SYS     SYS     NODE    NODE     X      PXB     NODE    SYS     SYS     SYS     SYS     SYS
NIC3    NODE    NODE    PXB     PIX     SYS     SYS     SYS     SYS     NODE    NODE    PXB      X      NODE    SYS     SYS     SYS     SYS     SYS
NIC4    NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE     X      SYS     SYS     SYS     SYS     SYS
NIC5    SYS     SYS     SYS     SYS     PIX     PXB     NODE    NODE    SYS     SYS     SYS     SYS     SYS      X      PXB     NODE    NODE    NODE
NIC6    SYS     SYS     SYS     SYS     PXB     PIX     NODE    NODE    SYS     SYS     SYS     SYS     SYS     PXB      X      NODE    NODE    NODE
NIC7    SYS     SYS     SYS     SYS     NODE    NODE    PIX     PXB     SYS     SYS     SYS     SYS     SYS     NODE    NODE     X      PXB     NODE
NIC8    SYS     SYS     SYS     SYS     NODE    NODE    PXB     PIX     SYS     SYS     SYS     SYS     SYS     NODE    NODE    PXB      X      NODE
NIC9    SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE    SYS     SYS     SYS     SYS     SYS     NODE    NODE    NODE    NODE     X 

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_0
  NIC1: mlx5_1
  NIC2: mlx5_2
  NIC3: mlx5_3
  NIC4: mlx5_4
  NIC5: mlx5_5
  NIC6: mlx5_6
  NIC7: mlx5_7
  NIC8: mlx5_8
  NIC9: mlx5_9

ulimit soft: 1048576
